### DISCUSSION TITLE
block editor : how many blocks are there ? And/or a way to test this ?

### DISCUSSION THREAD

**Discussion by manes6969** [tabemann/zeptoforth#118] [General]
Hi ,
Does anyone know how many blocks are available on a rpipico when using the block editor ?
Is there a way to know what the last available block is ?

Regards,
Ronny

**Comment by tabemann** [comment]
On Wed, Jul 24, 2024 at 4:24 AM manes6969 ***@***.***> wrote:

> Hi ,
> Does anyone know how many blocks are available on a rpipico when using the
> block editor ?
>

I should note that block ID's are not fixed indexes into arrays but rather
keys into a map; the only block ID that has a special value is $FFFFFFFF,
which should not be used. However, as for the total number of blocks that
can be used, irrespective of their ID's, if I recall correctly the number
is block::block-internal::sector-count 1-
block::block-internal::sector-block-count * or, on an rp2040_big build, 441
blocks

Is there a way to know what the last available block is ?
>

There really is no such thing as a last available block, because block ID's
are sparse. As for finding out the maximum block ID currently used, that is
not implemented as of yet.


> Regards,
> Ronny
>

Travis

**Comment by manes6969** [comment]
Hi Travis ,

Going further on this ; say I have used 50 blocks with code and running it , is there then a chance that one of the blocks could be overwritten , or is this secured by zeptoforth ? 

Regards,
Ronny

**Comment by tabemann** [comment]
On Wed, Jul 24, 2024 at 12:58 PM manes6969 ***@***.***> wrote:

> Hi Travis ,
>
> Going further on this ; say I have used 50 blocks with code and running it
> , is there then a chance that one of the blocks could be overwritten , or
> is this secured by zeptoforth ?
>
> Regards,
> Ronny
>

Blocks in zeptoforth are read-write; they can arbitrarily be both read and
written to provided the total number of blocks used is within the absolute
limit imposed by the flash storage for blocks. However, blocks will not be
written to provided you do not explicitly write to them (unless you are
using blocks for a FAT32 filesystem, where any operation that may result in
a write may modify any particular block).

Travis
